### PR TITLE
Added a Meaningful error msg for control server port taken

### PR DESCRIPTION
**Pull request by dhavalsingh**
### Description
Please describe your pull request. Thank you for contributing! You're the best.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

Closes [issue 3204](https://github.com/puma/puma/issues/3142)
Might close issue #2894 too.

### CODE REVIEW

File: test/config/activate_control_app.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+activate_control_app "tcp://0.0.0.0:9393"
+
+app do |env|
+  [200, {}, ["OK"]]
+end
```
Comment (dentarg) [inline test/config/activate_control_app.rb:5]: Why was this file added? Can't see in the diff that it is being used

File: lib/puma/runner.rb [hunk 2]
```diff
@@ -75,7 +75,11 @@ def start_control
       control = Puma::Server.new app, nil,
         { min_threads: 0, max_threads: 1, queue_requests: false, log_writer: @log_writer }
 
-      control.binder.parse [str], nil, 'Starting control server'
+      begin
+        control.binder.parse [str], nil, 'Starting control server'
+      rescue Errno::EADDRINUSE => e
+        raise "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
+      end
```
Comment (dentarg) [inline lib/puma/runner.rb:82]: Should we also re-raise the actual error? 

We could rescue all errors, add our helpful message and re-raise. I think that would cover https://github.com/puma/puma/issues/2894 too (seems like the exception raised on Windows is a different one)

Comment (ZaneBliss) [reply lib/puma/runner.rb:82]: @dentarg are you referring to something like this?

```ruby
begin
  control.binder.parse [str], nil, 'Starting control server'
rescue Errno::EADDRINUSE, Errno::EACCES => e
  raise e, "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
end
```

Also do you think there is some value in creating a descriptive exception class as well? 

(Perhaps `StandardError` isn't the right exception class to inherit from?)
```ruby
class ControlServerAddressPortConflict < StandardError; end
```

```ruby
begin
  control.binder.parse [str], nil, 'Starting control server'
rescue Errno::EADDRINUSE, Errno::EACCES => e
  raise ControlServerAddressPortConflict, "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
end
```

Comment (dhavalsingh) [reply lib/puma/runner.rb:82]: Now it should cover #2894 too i think.

Comment (ZaneBliss) [reply lib/puma/runner.rb:82]: Actually, after thinking about this I'm not so sure that it would handle #2894. Not that it couldn't but I would imagine the error message would at least need to be different. Sounds like @MSP-Greg started a branch for this last month but hasn't posted yet. I'd be interested in seeing what that looks like

Comment (dhavalsingh) [reply lib/puma/runner.rb:82]: yeah, might need to make the error msg a bit generic in that case.

Comment (dhavalsingh) [reply lib/puma/runner.rb:82]: Also, i think 
`class ControlServerAddressPortConflict < StandardError; end`
The error should be abstracted out in case we are raising Control server port issues in multiple (more than 1?) places. Otherwise doesn't make much sense IMO. Leme know what you guys think

Comment (ZaneBliss) [reply lib/puma/runner.rb:82]: I see where you're coming from with the abstraction comment, but I think it's a good idea because it will lead to a more helpful exception message stack trace. Even if it isn't used other places, I think that wrapping the exception in more domain specific terminology can help increase clarity imo 🙂. But it's your PR so use your discretion

Comment (nateberkopec) [reply lib/puma/runner.rb:82]: For me, Errno::EADDRINUSE, Errno::EACCES is clear enough 🤷 

I think we should rescue `Errno::EADDRINUSE, Errno::EACCES` here, as StandardError is too broad and may display this message in confusing circumstances.

**Review by nateberkopec — CHANGES_REQUESTED** [review]
See comment - change what class we're rescuing.

File: lib/puma/runner.rb (hunk 2 above)
Comment (nateberkopec) [reply lib/puma/runner.rb:82]: And as dentarg said, we should re-raise.

Comment (dhavalsingh) [reply lib/puma/runner.rb:82]: Fixed, re-raised the original error too.

**Comment by dhavalsingh** [comment]
Also, this branch is out-of-date with the base branch
How do I update this branch? With a rebase or a merge commit?

**Comment by dentarg** [comment]
You can do either, but I would recommend squashing the commits if you go for rebase. When merged, everything will be squashed anyway.

**Comment by dhavalsingh** [comment]
@nateberkopec this can be merged now?

**Comment by nateberkopec** [comment]
Thank you @dhavalsingh !
